src/python/pants/engine/native.py

```diff
@@ -216,6 +216,8 @@
 PyResult  execution_add_root_select(Scheduler*, ExecutionRequest*, Key, TypeConstraint);
 RawNodes* execution_execute(Scheduler*, ExecutionRequest*);
 
+PyResult capture_snapshot(Scheduler*, char*, Value);
```

Comment:
In general, for synchronous methods with concurrent/async implementations, I think it's a good idea to expose batch interfaces by default, as that sends the right signal about usage (that a caller should attempt to batch their calls wherever possible). Would recommend making this a batch method.

Reply:
I understand the desire, but will push back a little for now for two reasons:
1. I don't expect people to actually call this much at all, it's really just a short-term hack; if the do see people calling this, we need to design a better API.
2. The overhead of taking a list of things, which I think to do nicely would be: Create a new intrinsic type for `PathGlobsAndRoot`, pass it through everywhere, take a pointer to them with separate length, project out each thing from the tuple, seems like overkill. Happy to do this if we end up actually needing it, but again, if we find ourselves in that place we should probably design an actual API.

Reply:
> I don't expect people to actually call this much at all, it's really just a short-term hack; if the do see people calling this, we need to design a better API.

It will be used until "all" of pants is ported to the v2 engine in order to transition back and forth between local files and snapshots. And we immediately have a need for it to be concurrent I think, in order to capture ~500 jars into 500 snapshots on every `jvm_compile` startup. See https://github.com/pantsbuild/pants/issues/5762#issuecomment-385002308

> The overhead of taking a list of things, which I think to do nicely would be: Create a new intrinsic type for PathGlobsAndRoot, pass it through everywhere, take a pointer to them with separate length, project out each thing from the tuple, seems like overkill. 

I don't think you need a new intrinsic type, although wrapper `datatypes` would be useful. It would look like:
1. `project_multi` for a field that contains a list of PathGlobsAndRoots
2. project str for the root
3. use existing PathGlobs listing code for the rest

Reply:
Before jumping into discussion here: This is now done.

>> I don't expect people to actually call this much at all, it's really just a short-term hack; if the do see people calling this, we need to design a better API.
>
> It will be used until "all" of pants is ported to the v2 engine in order to transition back and forth between local files and snapshots.

I was expecting this to be a stop-gap until `BinUtils` is migrated into v2, and somehow exposed to v1 tasks...

> And we immediately have a need for it to be concurrent I think, in order to capture ~500 jars into 500 snapshots on every jvm_compile startup. See #5762 (comment)

I think I maybe don't understand #5762 then... What are those 500 jars? My understanding is that we _already_ create snapshots for targets when we parse BUILD files for them (that's why `./pants list ::` fills up `~/.cache/pants/lmdb_store`), but that we throw away the references to the Snapshots when we create the v1 Target type. I thought that #5762 was basically a "don't throw away those references" task, plus some "clean up some legacy stuff which means we have some convoluted paths" task, rather than a "do all of those snapshots twice" kind of task...?

Reply:
> I was expecting this to be a stop-gap until BinUtils is migrated into v2, and somehow exposed to v1 tasks...

There are lots of existing tasks that download things that will need to be ported/bootstrapped into the new model in order to be remoted... maybe this is the last integration that we will do starting from "mid graph", and all other task porting will be bottom up (and thus run resolvers in the datacenter with network access...?), but I have been assuming that ivy/coursier/npm/yarn/go/etc would need bootstrapping into Snapshots for now.

Reply:
> I think I maybe don't understand #5762 then... What are those 500 jars? 

The jars fetched by ivy and coursier: the 3rdparty dependencies.